### PR TITLE
Fixed DictFile.append(), Glidein logging, and added logserver

### PR DESCRIPTION
**Pull request by mambelli**
Fixed and disabled DictFile.append(), replaced by add_environment()

append was used only for environment variables in the condor submit file and was causing a malformed submit file.
add_environment was already used at times and is a better method. append was fixed and is still there commented in case we'd like to use it for other attributes in the future.

### CODE REVIEW

File: logserver/web-area/put.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+// Hardcoded parameters
+$configFile = '../logging_config.json';
+
+// Read and extract configuration
+$config = json_decode(file_get_contents($configFile), true);
+if (! $config) {
+    trigger_error("Empty config ($configFile) - the file may be missing or be invalid JSON", E_USER_WARNING);
+}
+$secretKeyFile = $config['secret_key_path'] ?? '/var/lib/gwms-factory/server-credentials/jwt_secret.key';
+$defaultSecretKey = $config['secret_key'] ?? 'VG9rZVNlY3JldEtleQo=';
```
Comment (namrathaurs) [inline logserver/web-area/put.php:43]: Is it okay/safe to have this secret key defined here (the actual key contents I meant)?

File: creation/web_base/logging_utils.source [hunk 2]
```diff
@@ -75,7 +77,7 @@ generate_glidein_metadata_json() {
 }
 
 
-log_init_tokens() {
+_glog_log_init_tokens() {
```
Comment (namrathaurs) [inline creation/web_base/logging_utils.source:80]: Should this be `_glog_init_tokens()`? 

If not, then lines 89, 94, and 99 should say `_glog_log_init_tokens`, right?

File: creation/web_base/logging_utils.source [hunk 3]
```diff
@@ -162,13 +164,13 @@ log_init() {
...
-    log_recipients=($(gconfig_get GLIDEIN_LOG_RECIPIENTS "${glidein_config}"))
+    log_recipients=($(gconfig_get GLIDEIN_LOG_RECIPIENTS_FACTORY "${glidein_config}") $(gconfig_get GLIDEIN_LOG_RECIPIENTS "${glidein_config}"))
     local no_send=0
     if [ "${#log_recipients[@]}" -eq 0 ]; then
-        warn "log_init: no recipients configured. Logs will still be produced, but not forwarded to remote servers."
+        warn "glog_init: no recipients configured. Logs will still be produced, but not forwarded to remote servers."
         no_send=1
-    elif ! log_init_tokens "${log_recipients[@]}"; then
-        warn "log_init: error while initializing tokens. Logs will still be produced, but not forwarded to remote servers."
+    elif ! _glog_log_init_tokens "${log_recipients[@]}"; then
```
Comment (namrathaurs) [inline creation/web_base/logging_utils.source:172]: Depending on what your answer is for my comment on line 80, this needs to be changed accordingly.

File: logserver/jwt.php [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+// SPDX-FileCopyrightText: 2009 Fermi Research Alliance, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+/*  Script to test JWT generation and verification in PHP:
+    php jwt.php // To run self test
+    php jwt.php TOKEN  // To decode the token
+    php jwt.php PAYLOAD TOKEN  // To encode PAYLOAD and save it to TOKEN
+    Uses the configuration in logging_config.json:
+    - secret_key_path (/var/lib/gwms-factory/server-credentials/jwt_secret.key)
+    - secret_key (VG9rZVNlY3JldEtleQo=)
```
Comment (namrathaurs) [inline logserver/jwt.php:11]: Is it okay/safe to have this secret key spelled out here?

File: logserver/jwt.php [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+// Hardcoded parameters
+$configFile = 'logging_config.json';
+// Read and extract configuration
+$config = json_decode(file_get_contents($configFile), true);
+if (! $config) {
+    trigger_error("Empty config ($configFile) - the file may be missing or be invalid JSON", E_USER_WARNING);
+}
+$secretKeyFile = $config['secret_key_path'] ?? '/var/lib/gwms-factory/server-credentials/jwt_secret.key';
+$defaultSecretKey = $config['secret_key'] ?? 'VG9rZVNlY3JldEtleQo=';
```
Comment (namrathaurs) [inline logserver/jwt.php:28]: Is it okay/safe to have this secret key defined here (the actual key itself)?

**Review by namrathaurs — COMMENTED** [review]
Some minor comments require to be addressed.

File: logserver/jwt.php (hunk 5 above)
Comment (mambelli) [reply logserver/jwt.php:28]: removed and protected from empty string

File: logserver/jwt.php (hunk 4 above)
Comment (mambelli) [reply logserver/jwt.php:11]: removed

File: logserver/web-area/put.php (hunk 1 above)
Comment (mambelli) [reply logserver/web-area/put.php:43]: removed and protected from empty string

File: creation/web_base/logging_utils.source (hunk 2 above)
Comment (mambelli) [reply creation/web_base/logging_utils.source]: is `_glog_init_tokens`, fixed

File: creation/web_base/logging_utils.source (hunk 3 above)
Comment (mambelli) [reply creation/web_base/logging_utils.source]: is `_glog_init_tokens`, fixed

File: CHANGELOG.md [hunk 6]
```diff
@@ -11,6 +11,7 @@ Changes since the last release
 
 -   item one of the list
 -   item N
```
[the remark this answers is not in the document]
Comment (mambelli) [reply CHANGELOG.md]: I did not want to change extra lines. Anyway fixed. Thanks

**Review by namrathaurs — DISMISSED** [review]
Changes look good and can be merged!

**Comment by mambelli** [comment]
@namrathaurs Sorry for the after-review changes. Testing the RPM installed version I noticed that there were some missing parts (only in my manual deployment) and I improved also the error checking.

File: logserver/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+dnf install php
+dnf install php-fpm
+systemctl start php-fpm
+systemctl enable php-fpm httpd
+```
+
+Ref: https://docs.redhat.com/en/documentation/red_hat_enterprise_linux/9/html/installing_and_using_dynamic_programming_languages/assembly_using-the-php-scripting-language_installing-and-using-dynamic-programming-languages
+Both PHP scripts use [Firebase PHP-JWT](https://github.com/firebase/php-jwt)
+installed via [Composer](https://getcomposer.org/download/)
+as done i n[this tutorial](https://www.sitepoint.com/php-authorization-jwt-json-web-tokens/).
```
Comment (namrathaurs) [inline logserver/README.md:30]: `i n[...` to be changed to `in [...`

**Review by namrathaurs — DISMISSED** [review]
Minor comments need to be addressed.

File: doc/factory/custom_scripts.html [hunk 8]
```diff
@@ -448,7 +449,32 @@ <h2><a name="xml_output"></a>Reporting script exit status</h2>
...
+        <p>
+          Standard output and standard error of all custom scripts (except the
+          periodic ones) is captured in the Glidein stdout and stderr and it is
+          transferred back to the Factory by HTCondor at the end of the Glidein.
+          Anyway this process may be insufficient: if the Glidein is killed the
+          transfer may not happen, if there are multi-Glideins all their
+          stdout/err are intermixed in the same files, and a user may desire to
+          have this output back earlier or in a different place. For all these
+          needs there is also a logging utility. It is defined in
+          <tt>logging_util.source</tt>, can be used in any custom script,
```
[the remark this answers is not in the document]
Comment (mambelli) [reply doc/factory/custom_scripts.html]: yes, fixed

File: doc/factory/custom_scripts.html [hunk 9]
```diff
@@ -448,7 +449,32 @@ <h2><a name="xml_output"></a>Reporting script exit status</h2>
...
+          set up <tt>GLIDEIN_LOG_RECIPIENTS_FACTORY</tt> as <tt>attr</tt> in the
+          Factory configuration. The Web servers at the URLs in
+          GLIDEIN_LOG_RECIPIENTS_FACTORY must be able to receive
+          JWT-authenticated PUT requests HS256-encoded with the secret set in
+          the Factory secret file
+          (<tt>/var/lib/gwms-factory/server-credentials/jwt_secret.key</tt>).
+          This secret file must be HMAC 256 compatible, e.g. a 32 bytes string.
+          The Factory will create the file at startup if it is not there or
+          empty. Scripts can use <tt>glog_setup</tt>, <tt>glog_write</tt>, and
+          <tt>glog_send_to_remote</tt>. There is an example of how to use
```
[the remark this answers is not in the document]
Comment (mambelli) [reply doc/factory/custom_scripts.html]: yes, fixed

File: logserver/README.md (hunk 7 above)
Comment (mambelli) [reply logserver/README.md]: yes, fixed

**Review by namrathaurs — APPROVED** [review]
Changes are ready for merging!

**Pull request merged**